### PR TITLE
[Model] GAT update

### PR DESCRIPTION
**Pull request by sufeidechabei**
I find pygat's accuracy is not consistent with the original paper, I have fixed that.
dgl cora    84.1%    original paper: 83.0 ± 0.7%
dgl citeseer 70.10% original paper: 72.5 ± 0.7%
dgl pubmed 77.6%  original paper: 79.0 ± 0.3%

### CODE REVIEW

**Comment by sufeidechabei** [comment]
Official code only opensource the training details of cora.  I will take some measures to improve the accuracy on citeseer and pubmed.

**Comment by jermainewang** [comment]
@yzh119 I remembered you had a smart implementation of multi-head attention in Transformer. Could it be applied to GAT too?

File: examples/pytorch/gat/gat.py [hunk 1]
```diff
@@ -26,14 +28,15 @@ def __init__(self, attn_drop):
...
-        ft = nodes.mailbox['ft'] # shape (B, deg, D)
+        a2 = nodes.mailbox['a2']  # shape (B, deg, 1)
+        ft = nodes.mailbox['ft']  # shape (B, deg, D)
         # attention
         a = a1 + a2  # shape (B, deg, 1)
         e = F.softmax(F.leaky_relu(a), dim=1)
         if self.attn_drop != 0.0:
-            e = F.dropout(e, self.attn_drop)
-        return {'accum' : torch.sum(e * ft, dim=1)} # shape (B, D)
+            e = F.dropout(e, self.attn_drop, training=self.training)
```
Comment (ylfdq1118) [inline examples/pytorch/gat/gat.py]: Instead of using nn.functional.dropout, let's change to a dropout layer, then we don't need to explicitly have `training=self.training`

File: examples/pytorch/gat/gat.py [hunk 2]
```diff
@@ -44,7 +47,8 @@ def __init__(self, headid, indim, hiddendim, activation, residual):
         self.residual_fc = None
         if residual:
             if indim != hiddendim:
-                self.residual_fc = nn.Linear(indim, hiddendim)
+                self.residual_fc = nn.Linear(indim, hiddendim, bias=False)
+                nn.init.xavier_normal_(self.residual_fc.weight.data, gain=1.414)
```
Comment (ylfdq1118) [inline examples/pytorch/gat/gat.py]: Would it better to change to `gain=nn.init.calculate_gain('relu'))`? which is equivalent to `gain=1.414`

File: examples/pytorch/gat/gat.py [hunk 3]
```diff
@@ -53,24 +57,32 @@ def forward(self, nodes):
                 ret = self.residual_fc(nodes.data['h']) + ret
             else:
                 ret = nodes.data['h'] + ret
-        return {'head%d' % self.headid : self.activation(ret)}
+        return {'head%d' % self.headid
+                : self.activation(ret)}
```
Comment (ylfdq1118) [inline examples/pytorch/gat/gat.py]: merge the above two lines

File: examples/pytorch/gat/gat.py [hunk 4]
```diff
@@ -53,24 +57,32 @@ def forward(self, nodes):
...
+        nn.init.xavier_normal_(self.fc.weight.data, gain=1.414)
+        nn.init.xavier_normal_(self.attn_l.weight.data, gain=1.414)
+        nn.init.xavier_normal_(self.attn_r.weight.data, gain=1.414)
 
     def forward(self, feats):
         h = feats
+
         if self.drop != 0.0:
-            h = F.dropout(h, self.drop)
+            h = F.dropout(h, self.drop, training=self.training)
```
Comment (ylfdq1118) [inline examples/pytorch/gat/gat.py]: use dropoutlayer

File: examples/pytorch/gat/gat.py [hunk 5]
```diff
@@ -190,30 +221,43 @@ def main(args):
         if epoch >= 3:
             dur.append(time.time() - t0)
 
+        if epoch%100 == 0:
+            acc = evaluate(model, features, labels, test_mask)
```
Comment (ylfdq1118) [inline examples/pytorch/gat/gat.py]: You should use validation mask instead of test mask

File: examples/pytorch/gat/gat.py [hunk 6]
```diff
@@ -190,30 +221,43 @@ def main(args):
...
+            model.train()
+            print("Test Accuracy {:.4f}".format(acc))
```
Comment (ylfdq1118) [inline examples/pytorch/gat/gat.py]: change to validation accuracy

Comment (sufeidechabei) [reply examples/pytorch/gat/gat.py]: Done

**Comment by ylfdq1118** [comment]
@sufeidechabei For the dropout, please make it a dropout layer. Check out the code in GCN:
https://github.com/dmlc/dgl/blob/master/examples/pytorch/gcn/gcn_spmv.py#L33

**Comment by sufeidechabei** [comment]
Done @ylfdq1118

**Comment by ylfdq1118** [comment]
@sufeidechabei I can't see any changes...

**Comment by ylfdq1118** [comment]
Why is the weight initialization removed?

File: examples/pytorch/gat/gat.py [hunk 7]
```diff
@@ -53,24 +60,34 @@ def forward(self, nodes):
...
+        self.fc = nn.Linear(indim, hiddendim, bias=False)
+        if drop:
+            self.drop = nn.Dropout(drop)
+        else:
+            self.drop = 0
+        self.attn_l = nn.Linear(hiddendim, 1, bias=False)
+        self.attn_r = nn.Linear(hiddendim, 1, bias=False)
+        nn.init.xavier_normal_(self.fc.weight.data, gain=1.414)
+        nn.init.xavier_normal_(self.attn_l.weight.data, gain=1.414)
+        nn.init.xavier_normal_(self.attn_r.weight.data, gain=1.414)
```
Comment (ylfdq1118) [inline examples/pytorch/gat/gat.py]: I remember you changed `gain` previously... I think it's because you copy-pasted and overwrote those changes.

Please make sure other places don't have similar errors.
